core/src/main/scala/kafka/log/AbstractLazyIndex.scala

```diff
@@ -0,0 +1,85 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package kafka.log
+
+import java.io.File
+import java.util.concurrent.locks.ReentrantLock
+
+import kafka.utils.CoreUtils.inLock
+
+/**
+ * An abstract lazy index wrapper that provides capability of avoiding index initialization
+ * on construction.
+ *
+ * This defers the index initialization to the time it gets accessed
+ * so the cost of the heavy memory mapped operation gets amortized over time.
+ */
+abstract class AbstractLazyIndex[T <: AbstractIndex[_, _]](private val _file: File) {
+
+  @volatile private var lazyIndex: LazyIndex = UninitializedLazyIndex(_file)
+  private val lock = new ReentrantLock()
+
+  sealed trait LazyIndex {
+    def file: File
+  }
+
+  case class UninitializedLazyIndex(indexFile: File) extends LazyIndex {
+    override def file: File = indexFile
+  }
+
+  case class InitializedLazyIndex(index: T) extends LazyIndex {
+    override def file: File = index.file
+  }
+
+  def file: File = lazyIndex.file
+
+  def file_=(f: File): Unit = {
+    inLock(lock) {
```

Comment:
unfortunately we cannot simply move this setter logic into the trait/case classes as well, since then we'd need to declare `indexFile` as `@volatile var indexFile` in `UninitializedLazyIndex` case class. This won't work for scala 2.11 due to https://github.com/scala/bug/issues/8873

Reply:
We don't need them to be case classes though. A plain old class will do. Also, we should move the sealed trait and subclasses to a companion object since they don't seem to need access to the outer class.

Reply:
agreed, moved into companion object